src/components/AuthForm.tsx

```diff
@@ -0,0 +1,269 @@
+import React, { useEffect, useState } from "react";
+import {
+  getAuth,
+  signInWithEmailAndPassword,
+  createUserWithEmailAndPassword,
+} from "firebase/auth";
+import { app } from "../firebase";
+import { collection, addDoc } from "firebase/firestore";
+import { db } from "../firebase";
+import { object, string } from "yup";
+import { useAuth } from "../contexts/AuthContext";
+import classes from "../styles/AuthForm.module.scss";
+
+interface AuthFormProps {
+  isSignIn: boolean;
+  setIsSignIn: (value: boolean) => void;
+}
+
+interface AuthData {
+  username?: string;
+  email: string;
+  password: string;
+  confirmPassword?: string;
+}
+
+type FormError = {
+  fieldName: keyof AuthData;
+  message: string;
+};
+
+const AuthForm: React.FC<AuthFormProps> = ({ isSignIn, setIsSignIn }) => {
+  const [authData, setAuthData] = useState<AuthData>({
+    email: "",
+    password: "",
+  });
+
+  const [errors, setErrors] = useState<FormError[]>([]);
+  const [error, setError] = useState<string>("");
+
+  const { login } = useAuth();
+
+  useEffect(() => {
+    setAuthData({
+      email: "",
+      password: "",
+      username: "",
+      confirmPassword: "",
+    });
+    setErrors([]);
+    setError("");
+  }, [isSignIn]);
+
+  const handleInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    const { name, value } = e.target;
+    setAuthData({ ...authData, [name]: value });
+  };
+
+  const handleBlur = (e: React.FocusEvent<HTMLInputElement>) => {
+    const { name, value } = e.target;
+    validateField(name, value);
+  };
+
+  const validateField = (name: string, value: string) => {
+    try {
+      const fieldSchema = object().shape({
+        username:
+          !isSignIn && name === "username"
+            ? string().required("Username is required")
+            : string(),
+        email:
+          name === "email"
+            ? string().email("Invalid email").required("Email is required")
+            : string(),
+        password:
+          name === "password"
+            ? string()
+                .min(6, "Password must be at least 6 characters")
```

Comment:
all minor, but generally we would like to have named values. From this it is clear that 6 is a number of characters in the string, but we call this a magic variable. because you need to look at the sounding context to understand what is this value. but if you move this number up in constant and name it "PASSWORD_MIN_CHAR". This is a trivial example, but lets try to avoid magic variables in our coding practices